agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -231,32 +231,35 @@ impl DeriveSchema for Stage {
                     .split('.')
                     .map(|s| s.to_string())
                     .collect::<Vec<String>>();
-                macro_rules! get_path_schema {
-                    () => {{
-                        // Unfortunately, the borrow checker requires us to compute path_schema in
-                        // both branches because we have a mutable borrow on state when
-                        // we compute the fill_schema in the Value case. This macro cleans up some
-                        // duplication.
-                        get_schema_for_path_mut(&mut state.result_set_schema, path_vec)
-                            .ok_or_else(|| Error::UnknownReference(path.clone()))?
-                    }};
-                }
                 match fill_output {
                     FillOutput::Value(e) => {
                         let fill_schema = e.derive_schema(state)?;
-                        let path_schema = get_path_schema!();
-                        // Null is possible if the value expression can evaluate to null, but
-                        // missing will always be impossible. We subtract NULL so it can only be
-                        // reintroduced via the union of the fill_schema
-                        *path_schema = std::mem::take(path_schema)
-                            .subtract_nullish()
-                            .union(&fill_schema);
+                        match get_schema_for_path_mut(
```

Comment:
for $fill, we were correctly marking existing keys as required; but we were not inserting as required new keys.